File: src/main/java/cses/sorting_and_searching/FerrisWheel.java
package cses.sorting_and_searching;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;

public class FerrisWheel {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int x = Integer.parseInt(st.nextToken());

        List<Integer> weights = new ArrayList<>(n);
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            weights.add(Integer.parseInt(st.nextToken()));
        }

        Collections.sort(weights); // uses TimSort in Java 8

        int i = 0;
        int j = n - 1;
        int result = 0;

        while (i <= j) {
            if (weights.get(i) + weights.get(j) <= x) {
                i++;
            }
            j--;
            result++;
        }

        System.out.println(result);
    }
}

File: src/main/kotlin/cses/sorting_and_searching/FerrisWheel.kt
package cses.sorting_and_searching

fun main() {
//    val (n, x) = readln().split(" ").map { it.toInt() }
//    val childrenWeights = readln().split(" ").map { it.toInt() }
    val n = 4
    val x = 10
    val childrenWeights = listOf(7, 2, 3, 9)

    val answer = solution(n, x, childrenWeights)

    println(answer)
}

private fun solution(n: Int, x: Int, weights: List<Int>): Int {
    val sortedWeights = weights.sorted()
    var i = 0
    var j = n-1
    var result = 0

    while (true) {
        if (i==j) {
            result++
            break
        }
        if (sortedWeights[j] + sortedWeights[i] <= x) i++
        j--
        result++
    }

    return result
}
